Update for Entity Framework Core 3.1

nizmow opened the pull request:
Mostly fairly simple. I'm not that happy with the code for custom queries, but I didn't work too hard as it might be a good idea to pull it, since the other providers don't support it and there are other ways of loading related data.

Tests are green.

Relates to  #1628

maldworth commented on src/Persistence/MassTransit.EntityFrameworkCoreIntegration/MassTransit.EntityFrameworkCoreIntegration.csproj:
```diff
@@ -14,7 +14,8 @@
 
   <ItemGroup>
-      <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="2.2.6" />
+      <PackageReference Include="Microsoft.Data.SqlClient" Version="1.1.0" />
```
Lets remove the explicit upgrade of Microsoft.Data.SqlClient, and instead let it resolve through the transitive dependency within Microsoft.EntityframeworkCore.  I always prefer to use the minimum, and let the user of the library upgrade if they want.

maldworth commented on src/Persistence/MassTransit.EntityFrameworkCoreIntegration/Saga/EntityFrameworkSagaRepository.cs:
```diff
@@ -143,15 +143,13 @@ await dbContext.Database.BeginTransactionAsync(_isolationLevel, context.Cancella
                     var rowLockQuery = _rawSqlLockStatements?.GetRowLockStatement<TSaga>(dbContext);
                     if (rowLockQuery != null)
-                        instance = await queryable.FromSql(rowLockQuery, new object[] {sagaId}).SingleOrDefaultAsync(context.CancellationToken)
+                        instance = await ApplyCustomQuery(dbContext, dbSet => dbSet.FromSqlRaw(rowLockQuery, new object[] {sagaId})).SingleOrDefaultAsync()
```
don't forget the  `.SingleOrDefaultAsync(context.CancellationToken).ConfigureAwait(false)`

maldworth commented on src/Persistence/MassTransit.EntityFrameworkCoreIntegration/MsSqlLockStatements.cs:
```diff
@@ -26,18 +26,15 @@ public class MsSqlLockStatements : IRawSqlLockStatements
 
```
Damn github PR isn't letting me put a comment on an "unchanged" line. Anyways, Line 19 is a using statement that is unused. So it can be removed.

maldworth commented:
Minor tweaks, but PR looks good otherwise. Thanks for the submit!

NooNameR commented on src/Persistence/MassTransit.EntityFrameworkCoreIntegration/MassTransit.EntityFrameworkCoreIntegration.csproj:
```diff
@@ -14,7 +14,8 @@
   <ItemGroup>
-      <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="2.2.6" />
+      <PackageReference Include="Microsoft.Data.SqlClient" Version="1.1.0" />
+      <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="3.1.0" />
```
Why we need this reference?

nizmow replied:
There's SQL Server specific code in here. It would be a lot more work to make this database-agnostic (and relying on `Microsoft.EntityFrameworkCore.Relational`). If that's what you mean?

NooNameR replied:
right now EF Core lib is related to SQLServer but if I want PostgreSQL?
I know that was before.. But wondering if we could do it better
For eg, we could have table name inside `SagaClassMap` but if we have custom `DbContext` it won't work

maldworth replied:
Yeah, try bumping it to the Relational package, then it would be db-agnostic. The only thing you will need to do is update [this line](https://github.com/MassTransit/MassTransit/blob/b49e9ddc60c73cdf894cfc79a6c7f22a90da6b6b/src/Persistence/MassTransit.EntityFrameworkCoreIntegration/Saga/EntityFrameworkSagaRepository.cs#L104). There is probably a base type, you could use that instead. There's a unit test for that, so if it still passes, then that should be acceptable.

nizmow replied:
I had a look at the EF core, the default implementation returns a `NoopExecutionStrategy`, it appears we can just use that regardless and get rid of the condition.

nizmow replied:
There's another thing we'd need to fix to make it work with multiple backends, it seems there's code to [detect deadlocks](https://github.com/MassTransit/MassTransit/blob/b49e9ddc60c73cdf894cfc79a6c7f22a90da6b6b/src/Persistence/MassTransit.EntityFrameworkCoreIntegration/Saga/EntityFrameworkSagaRepository.cs#L407) which obviously uses `Microsoft.Data.SqlClient` directly. It seems to only be used to [avoid a transaction rollback](https://github.com/MassTransit/MassTransit/blob/b49e9ddc60c73cdf894cfc79a6c7f22a90da6b6b/src/Persistence/MassTransit.EntityFrameworkCoreIntegration/Saga/EntityFrameworkSagaRepository.cs#L357) in the case of a deadlock. I suppose this is for performance reasons. We have a few options:

1) Get rid of that code and rollback anyway.
2) Always return false if we don't have a `SqlException`, so the detection will only work on SQL Server.
3) Leave the code SQL Server specific.

Ideas?

maldworth replied:
I think if we look for the type `ExecutionStrategy` instead, then that [should work](https://github.com/aspnet/EntityFrameworkCore/blob/master/src/EFCore.SqlServer/SqlServerRetryingExecutionStrategy.cs).

I see the factory you are looking at here with [Noop](https://github.com/aspnet/EntityFrameworkCore/blob/master/src/EFCore.Relational/Storage/RelationalExecutionStrategyFactory.cs). So if NoopExecutionStrategy is returned, `if (execStrategy isExecutionStrategy)` should skip and process normally.

nizmow replied:
Pushed database agnostic implementation. Removed checks for deadlocks that avoided rollbacks, if anyone has a good reason for them to remain there then I think we'll have to stick to MSSQL-only and revert this commit. Cleaned up some exception handling in the process.

Tests green.

maldworth replied:
Regarding the Deadlock detection. I'm not sure the correct answer, but my guess would be that the majority of people using EfCore are MsSql. So leaving the deadlock detection (and dependency on the `Microsoft.Data.SqlClient` package), is probably the best for the users.

We can always re-visit if we get complaints from people about this package being used unnecessarily.

nizmow replied:
I jumped the gun there!

nizmow commented:
Fixed suggested issues, waiting on more feedback from @NooNameR.

phatboyg commented:
So is this gtg? I don't want to merge it unless it's ready - since there may be minor fixes that need to go in a 6.0.1

NooNameR commented:
I'm still no sure about SQL related stuffs

NooNameR commented:
@maldworth why couldn't we use `Rollback` for all cases? if we have deadlock in `SQLServer` world it won't do nothing, I'm wrong? we could just wrap it in other try/catch

nizmow commented:
> @maldworth why couldn't we use `Rollback` for all cases? if we have deadlock in `SQLServer` world it won't do nothing, am I wrong? we could just wrap it in other try/catch

That's what I thought, I have no idea why that special behaviour is there except perhaps some kind of performance improvement (but SQL Server will rollback by itself on the deadlock victim). For simplicity I'd consider removing it altogether.

I should be able to put this into a package and run some real load over it within the next few days. Aside from the issue above I think it's ready for merge. If we want to be more conservative I think we should stick to supporting SQL Server only.

NooNameR commented:
I haven't been working with SQL for a long time, but isn't it saves to Db when you call `transaction.Commit()`

nizmow commented:
@NooNameR of course, unless the commit fails. But the strange deadlock code is trying to avoid an explicit rollback in the case of a deadlock. I don't think it's required, but I'm not even sure it's beneficial (I've tried to do some research). @maldworth seems to have more SQL Server experience than I, so perhaps he has some idea why it's there. If it were up to me I'd be happy to remove it.

NooNameR commented:
I think we should left rollback for any cases and no be worried about deadlock or not

maldworth commented:
So dont believe I added that deadlock code check to the original ef implementation. So if we want to nuke it in favor of eliminating the mssql specific dependency, then I'm fine.

phatboyg commented:
Can we elminate it, and have unit tests for both SQl and Postgres that verify behavior? Surely we can simulate a deadlock with a slow transaction running while trying to insert a saga or something.

nizmow commented:
Sure, I should be able to work on those tests over the next few days. Especially tests for a second provider would be a good idea, if we are going to support it. I'm blocked myself because of lack of support for EFCore 3.1, but I will fork this for internal use which will keep us going until it's properly ready (also a good opportunity for some battle testing).

nizmow commented:
I've done the work to add explicit support for PostgreSQL and done a fairly major overhaul of the tests to cover PostgreSQL. There's some small external API changes in the IRawSqlLockStatements implementations, previously there was a SQL Server lock statements class that had some virtual methods, though I'm unsure why. This is now a generic relational lock statements class with some derived classes for SQL Server and PSQL.

Added a contrived test to cause deadlocks to see how the deadlock handling worked and it was not effective so I've removed it.

nizmow commented:
Anyone have time to take another quick look at this? :)

maldworth commented:
After this weekend I'll take another look. Also hopefully there wont be any merge conflicts with @phatboyg saga configuration PR. But whatever one gets merged in last might need a rebase.

phatboyg commented:
I'm hoping to get to it this weekend. Forecast? Cold and rainy, so yeah, a good weekend to stay inside and knock out some code...

phatboyg commented:
So I think I have all this integrated in the other PR

nizmow commented:
Looks good, thanks!

phatboyg commented:
Really loving the Postgres support. That, combined with the docker compose to get SQL Server and Postgres running on my MacBook Pro is seriously "lit." More unit tests I can run, ftw. :)
